src/librustc_ty/ty.rs

```diff
@@ -367,6 +369,103 @@ fn asyncness(tcx: TyCtxt<'_>, def_id: DefId) -> hir::IsAsync {
     fn_like.asyncness()
 }
 
+/// For associated types we allow bounds written on the associated type
+/// (`type X: Trait`) to be used as candidates. We also allow the same bounds
+/// when desugared as bounds on the trait `where Self::X: Trait`.
+///
+/// Note that this filtering is done with the trait's identity substs to
+/// simplify checking that these bounds are met in impls. This means that
+/// a bound such as `for<'b> <Self as X<'b>>::U: Clone` can't be used, as in
+/// `hr-associated-type-bound-1.rs`.
+fn associated_type_projection_predicates(
+    tcx: TyCtxt<'_>,
+    def_id: DefId,
+) -> &'_ ty::List<ty::Predicate<'_>> {
+    let trait_id = tcx.associated_item(def_id).container.id();
+    let trait_substs = InternalSubsts::identity_for_item(tcx, trait_id);
+
+    let generic_trait_bounds = tcx.predicates_of(trait_id);
```

Comment:
This PR now does this feature gated on `generic_associated_type` because it changes a lot of tests and I don't really want to introduce soundness holes in a PR that's supposed to be a soundness fix.